lib/Dodawanie_kolumny_z_epokami.py

```diff
@@ -0,0 +1,19 @@
+from pyspark.sql.functions import col, when
+from pyspark.sql.dataframe import DataFrame
+
+def add_epoch_column(df: DataFrame) -> DataFrame:
+  periods = [1901,1918,1926,1939,1954,1970,1985,1994,2009]
+  df_no_N = df.filter(df.startYear != "\\N")
+
+  df_periods = df_no_N.withColumn('period',
+                             when(col('startYear') <= periods[0], "1")
+                             .when(col('startYear') <= periods[1], "2")
+                             .when(col('startYear') <= periods[2], "3")
+                             .when(col('startYear') <= periods[3], "4")
+                             .when(col('startYear') <= periods[4], "5")
+                             .when(col('startYear') <= periods[5], "6")
+                             .when(col('startYear') <= periods[6], "7")
+                             .when(col('startYear') <= periods[7], "8")
+                             .when(col('startYear') <= periods[8], "9")
```

Comment:
jest szansa, że tu zadziałała by taka formuła napisana reduce'm
```
formulas = [ when(col('startYear') <= period, str(index)+1)  for period, index in zip(periods, range(len(periods))]
formulas += [.otherwise( len(periods) +1)]
from itertools import reduce

f_periods = df_no_N.withColumn('period',
    reduce(lambda x,y: x.y, formulas)
    )  
```

Reply:
Niestety powyższa formuła nie działa, próbowałem ją modyfikować na różne sposoby, lecz również bez skutku.

Reply:
A mnie się udało:
```
from pyspark.sql.functions import col, when
from pyspark.sql import DataFrame
from functools import reduce


def add_epoch_column(df: DataFrame, periods = [1901,1918,1926,1939,1954,1970,1985,1994,2009]) -> DataFrame:

  assert df.filter(df.rok_wydania_produkcji == "\\N").count() == 0
  assert dict(df.dtypes)["rok_wydania_produkcji"] == "int"
 
  formulas = [ 
    when(col('startYear') <= period, str(index)+1)  
    for period, index 
    in zip(periods, range(len(periods)))]


  df_periods = df.withColumn('period',
    reduce(lambda x,y: x.y, formulas).otherwise( len(periods) +1)
  )  

  return df_periods
```

Reply:
Niestety wciąż jest problem z tą formułą. Otrzymuję taki błąd:

> Can't extract value from CASE WHEN (rok_wydania_produkcji#2180 <= 1901) THEN 1 END: need struct type but got string

Reply:
Test przechodzi. W jakiej sytuacji obserwuje Pan błąd?

Reply:
Problem jest już na początku z powodu `str(index)+1`. Po poprawie tego otrzymuję błąd 

> otherwise() can only be applied on a Column previously generated by when()

Zatem moim pomysłem jest, aby zamienić `otherwise()` na kolejne `when()`:

```
from pyspark.sql.functions import col, when
from pyspark.sql import DataFrame
from typing import Optional,List
from functools import reduce

DEFAULT_PERIODS = [1901,1918,1926,1939,1954,1970,1985,1994,2009]

def add_epoch_column(df: DataFrame, periods: Optional[List[int]] = None) -> DataFrame:

  if periods is None:
    periods = DEFAULT_PERIODS

  assert df.filter(df.rok_wydania_produkcji == "\\N").count() == 0
  assert dict(df.dtypes)["rok_wydania_produkcji"] == "int"

  formulas = [ 
    when(col('rok_wydania_produkcji') <= period, str(index+1))  
    for period, index 
    in zip(periods, range(len(periods)))]
  formulas.append(when(col("rok_wydania_produkcji") > periods[-1], str(len(periods) + 1)))

  df_periods = df.withColumn('epoka',
    reduce(lambda x,y: x.y, formulas)
  )  
   
  return df_periods
```
I w tym momencie otrzymuję wspomniany wyżej błąd.

Reply:
Chce zobaczyć, że moje rozwiazanie u Ciebie nie działa. Umów spotkanie Janie

Reply:
zmieniłem na formułę zaprezentowaną wczoraj